### PR TITLE
AOC-026: Day 9: Marble Mania

### PR DESCRIPTION
**Pull request by basseur**
This is WIP. Don't merge!

### CODE REVIEW

**Comment by basseur** [comment]
[AOC-026: Day 9: Marble Mania](https://trello.com/c/KKYeFZ1l/20-aoc-026-day-9-marble-mania)

File: src/main/java/org/basseur/adventofcode/advent2018/days/day09/CircleDeque.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.ArrayDeque;
+import java.util.Collection;
+
+/**
+ * Implementation of a circular ArrayDeque which adds a method to rotate the Deque.
+ *
+ * @param <E> the type of elements held in this collection
+ * @author Jan Philipp G&ouml;lz
+ */
+public class CircleDeque<E> extends ArrayDeque<E> {
```
Comment (mfbieber) [inline src/main/java/org/basseur/adventofcode/advent2018/days/day09/CircleDeque.java]: nice idea, I didn't know the ArrayDeque before :)

Comment (mfbieber) [reply src/main/java/org/basseur/adventofcode/advent2018/days/day09/CircleDeque.java]: How is the performance here? We used linked lists to solve this...Matthias implementation was rather slow (2-3 h), mine a little faster (< 10 min) but with a improved implementation - and yours?

Does the user have to wait quite some time to see the result? Now things get interesting. We should consider to come back to the user via email or so, if the calculation takes too long...This is why were thinking about persisting results and user accounts, etc :)

File: src/main/java/org/basseur/adventofcode/advent2018/days/day09/CircleDeque.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    /**
+     * Rotates this CircleDeque by the number of steps given as an argument.
+     * A positive value means clockwise, a negative value means counterclockwise.
+     * For this purpose, the last (resp. first) element is removed and added again as the first (resp. last) element
+     * depending on the direction of rotation.
+     *
+     * @param steps number of steps to rotate the circle. Positive is clockwise, negative is counterclockwise.
+     */
+    public void rotate(int steps) {
```
Comment (mfbieber) [inline src/main/java/org/basseur/adventofcode/advent2018/days/day09/CircleDeque.java]: cool. So you are implementing this as if it was a roulette or so? Really DDD here ;)

File: src/main/java/org/basseur/adventofcode/advent2018/days/day09/CircleDeque.java (hunk 1 above)
Comment (basseur) [reply src/main/java/org/basseur/adventofcode/advent2018/days/day09/CircleDeque.java]: I wanted to implement this as a circle and found information about doubly linked lists and circular doubly linked lists. Somewhere I found, that you could do that with Deque (pronounced deck), which is short for _double-ended queue_.

And it is extremely fast: For part 1 it takes about 1.8 ms, for part 2 it takes about 284.2 ms (on my machine)! :D
Try here: https://advent-of-code-app-pr-73.herokuapp.com/

**Pull request merged**